### PR TITLE
Align Datastructures to Spec and Add Fixed-Length Tuple Support

### PR DESCRIPTION
**Pull request by dangerousfood**
This PR rewrites the datastructures to conform to the v0.5.1 Beacon Chain spec, and also adds support for "tuples" (fixed-length lists), which are serialized without the length mixin.

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconState.java [hunk 1]
```diff
@@ -62,10 +62,12 @@
...
   // Recent state
+  // TODO This is bounded by SHARD_COUNT
   protected List<Crosslink> latest_crosslinks;
-  protected List<Bytes32> latest_block_roots;
-  protected List<Bytes32> latest_state_roots;
-  protected List<Bytes32> latest_active_index_roots;
+  protected List<Bytes32> latest_block_roots; // Bounded by SLOTS_PER_HISTORICAL_ROOT
+  protected List<Bytes32> latest_state_roots; // Bounded by SLOTS_PER_HISTORICAL_ROOT
+  protected List<Bytes32> latest_active_index_roots; // Bounded by LATEST_ACTIVE_INDEX_ROOTS_LENGTH
+  // TODO This is bounded by LATEST_SLASHED_EXIT_LENGTH
```
Comment (akhila-raju) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconState.java:70]: Why is this a TODO here?

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Fork.java [hunk 2]
```diff
@@ -20,7 +20,9 @@
 
 public class Fork {
 
+  // TODO previous_version should be a bytes4 (this has serialization impacts)
   private Bytes previous_version;
+  // TODO previous_version should be a bytes4 (this has serialization impacts)
```
Comment (akhila-raju) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Fork.java]: Please update this comment to current_version

**Review by akhila-raju — COMMENTED** [review]
1 comment, otherwise LGTM

**Comment by jrhea** [comment]
I have proven that steven's PR is writable.  I think etiquette dictates that you should just push you last fix to his PR and close this.  I have posted instructions on how to do it in slack

**Pull request closed without merging**